### PR TITLE
[MOSS] Decouple MOSS-TTS Local codec from processor

### PR DESCRIPTION
**Pull request by Ratish1**
## Motivation

MOSS-TTS Local preprocessing and vocoder stages previously reached through the Hugging Face processor-owned `audio_tokenizer`. This keeps the codec load and processor state coupled, which makes the local codec path harder to reason about and harder to optimize independently.

This PR loads the MOSS-Audio-Tokenizer-v2 codec as an owned component for MOSS-TTS Local while keeping the processor responsible for text/config handling.

## Modifications

- Add an owned `MossTTSLocalAudioTokenizer` wrapper that loads `OpenMOSS-Team/MOSS-Audio-Tokenizer-v2`, prepares reference waveforms with the upstream channel/resample/loudness semantics, and exposes a narrow encode API.
- Load the MOSS-TTS Local processor without constructing its codec, then load the codec separately for preprocessing and vocoder stages.
- Route reference file paths and data-URI references through the owned codec encoder with the existing batching/cache path.
- Pass the owned codec model directly into the streaming vocoder scheduler instead of deriving it from `processor.audio_tokenizer`.
- Add unit coverage for owned codec loading, waveform preparation, batched reference encoding, data-URI handling, and vocoder executor construction.

## Validation

Unit gate:

```text
115 passed, 20 warnings
```

Full SeedTTS EN, `--max-concurrency 8`:

| metric | value |
|---|---:|
| completed / failed | `1088 / 0` |
| WER evaluated / skipped | `1088 / 0` |
| throughput_qps | `7.138` |
| rtf_mean / p95 / p99 | `0.2637 / 0.3794 / 0.4389` |
| latency mean / p95 / p99 | `1.119s / 1.530s / 1.717s` |
| output throughput | `392.5 tok/s` |
| output tokens total | `59832` |
| WER corpus | `2.28%` |
| WER below 50 | `1.83%` |
| >50% WER outliers | `5` |
| UTMOS mean | `3.9542` |
| speaker similarity mean | `65.1585` |

Audio integrity:

| check | value |
|---|---:|
| files | `1088` |
| missing | `0` |
| read errors | `0` |
| silent / zero-peak | `0` |
| empty | `0` |
| sample rate | `48000` for all |
| channels | stereo for all |
| duration min / mean / max | `1.68s / 4.399s / 9.36s` |
| full-scale clipped sample count | `20` |

### CODE REVIEW

File: sglang_omni/models/moss_tts_local/streaming_vocoder.py [hunk 1]
```diff
@@ -848,8 +833,6 @@ def _decode_codes_rows(self, codes_list: list[torch.Tensor]) -> list[torch.Tenso
             # The processor helper forces chunk_duration=8 and enters the
             # tokenizer streaming loop. This decoder is non-streaming, so it must
             # run through the tokenizer's full-sequence decode path.
-            # TODO(Ratish): Load and own the non-streaming codec directly so this
-            # path does not need to swap the processor-owned decoder at call time.
             original_decoder = self._codec.decoder
```
Comment (JiaxinD) [inline sglang_omni/models/moss_tts_local/streaming_vocoder.py:836]: Confirmed: the #822 decoder swap and the #798 graph capture both operate on `self._codec` here, which is now the directly-owned codec, so the mechanism is unchanged and I verified there is no break. Removing the TODO is right since the ownership half landed; the swap staying is expected (one codec, two decode modes).

File: sglang_omni/models/moss_tts_local/audio_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                "audio_codes/audio_codes_lengths"
+            )
+        codes_cpu = audio_codes.detach().to("cpu", torch.long)
+        lengths_cpu = audio_codes_lengths.detach().to("cpu")
+        return [
+            codes_cpu[:, index, : int(lengths_cpu[index])].transpose(0, 1).contiguous()
+            for index in range(int(codes_cpu.shape[1]))
+        ]
+
+    def _prepare_waveform(self, wav: torch.Tensor, sample_rate: int) -> torch.Tensor:
```
Comment (JiaxinD) [inline sglang_omni/models/moss_tts_local/audio_tokenizer.py:110]: Small ordering asymmetry: path jobs resample in `load_paths` before this folds channels, while wav/data-URI jobs fold here then resample. It commutes for the repeat/truncate cases so output matches, but it is a latent footgun, and `test_audio_tokenizer_path_resamples_before_channel_fold` feeds stereo input so the fold is a no-op there (it does not actually exercise resample-before-fold).

File: sglang_omni/models/moss_tts_local/stages.py [hunk 3]
```diff
@@ -488,9 +580,14 @@ def create_preprocessing_executor(
...
-    processor = _load_moss_tts_local_processor(model_path, device=device)
+    processor = _load_moss_tts_local_processor(model_path)
+    audio_tokenizer = load_moss_tts_local_audio_tokenizer(
+        _resolve_audio_tokenizer_model_path(processor, codec_model_path),
+        device=device,
+    )
     reference_encoder: Any = _BatchedReferenceEncoder(
-        processor,
+        audio_tokenizer,
+        n_vq=int(processor.model_config.n_vq),
```
Comment (JiaxinD) [inline sglang_omni/models/moss_tts_local/stages.py:590]: `num_quantizers` is now passed explicitly as `model_config.n_vq`, whereas the old `processor.encode_audios_from_path` took no such arg. Worth confirming `n_vq` matches the codec's default quantizer count so reference codes stay identical to the previous path.

File: sglang_omni/models/moss_tts_local/audio_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        f"Loading MOSS-TTS Local audio tokenizer from {checkpoint_dir} on {device}"
+    )
+    try:
+        from transformers import AutoModel
+
+        with moss_transformers_processor_compat():
+            model = AutoModel.from_pretrained(
+                checkpoint_dir,
+                trust_remote_code=True,
+            )
```
Comment (SandyLuXY) [inline sglang_omni/models/moss_tts_local/audio_tokenizer.py:160]: The owned MOSS-TTS Local codec loader calls `AutoModel.from_pretrained(...)` without passing `codec_weight_dtype="bf16"`. The old processor-owned loading path defaulted the codec weights to bf16 before loading `MOSS-Audio-Tokenizer-v2`; without that kwarg, the codec model falls back to fp32 when its config does not provide `codec_weight_dtype`.

Comment (SandyLuXY) [reply sglang_omni/models/moss_tts_local/audio_tokenizer.py:160]: Please check line 216 of https://huggingface.co/OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5/blob/main/processing_moss_tts.py

File: sglang_omni/models/moss_tts_local/stages.py (hunk 3 above)
Comment (Ratish1) [reply sglang_omni/models/moss_tts_local/stages.py:590]: Verified this, HF _assert_fixed_nq(None) resolves to model_config.n_vq, and it shows same-mode upstream/candidate reference codes are exact.

File: sglang_omni/models/moss_tts_local/audio_tokenizer.py (hunk 4 above)
Comment (Ratish1) [reply sglang_omni/models/moss_tts_local/audio_tokenizer.py:160]: Good catch. Fixed by passing codec_weight_dtype="bf16" to AutoModel.from_pretrained, matching HF processing_moss_tts.py.

File: sglang_omni/models/moss_tts_local/audio_tokenizer.py (hunk 2 above)
Comment (Ratish1) [reply sglang_omni/models/moss_tts_local/audio_tokenizer.py:110]: Done. I checked the current HF processor: `encode_audios_from_path` resamples before delegating to `encode_audios_from_wav`, while direct wav/data-URI encoding folds before resample. Kept that behavior and changed the test to use mono input so it actually verifies path resample happens before channel fold.

File: sglang_omni/models/moss_tts_local/stages.py [hunk 5]
```diff
@@ -62,6 +70,40 @@ class _ArMemoryBudget:
...
+    audio, sample_rate = sf.read(io.BytesIO(raw), dtype="float32", always_2d=True)
+    duration = audio.shape[0] / max(int(sample_rate), 1)
+    if duration > _MAX_REFERENCE_SECONDS:
+        raise ValueError(
+            f"reference audio is {duration:.1f}s long, limit is "
+            f"{_MAX_REFERENCE_SECONDS:.0f}s"
+        )
+    return torch.from_numpy(audio.T), int(sample_rate)
+
+
```
Comment (Ccyest) [inline sglang_omni/models/moss_tts_local/stages.py]: Nit: Looks like these two functions are helpers extracted from `CachedReferenceEncoder.encode_data_uri`, now placed as module-level functions at the top of stages.py. 

Consider moving these two functions into `audio_tokenizer.py` or making them @staticmethod on _BatchedReferenceEncoder to reduce stages.py bulk

File: sglang_omni/models/moss_tts_local/stages.py [hunk 6]
```diff
@@ -142,34 +184,50 @@ def _resolve_codec_device(device: str | None, gpu_id: int | None) -> str:
...
+) -> str:
+    if codec_model_path is not None:
+        return codec_model_path
+    return str(
+        getattr(
+            processor.model_config,
+            "audio_tokenizer_name_or_path",
+            DEFAULT_MOSS_TTS_LOCAL_AUDIO_TOKENIZER,
+        )
+        or DEFAULT_MOSS_TTS_LOCAL_AUDIO_TOKENIZER
```
Comment (Ccyest) [inline sglang_omni/models/moss_tts_local/stages.py]: Nit: getattr function should already have a default value, wondering why we have a "or" guarding here?

**Review by Ccyest — APPROVED** [review]
lgtm, just some nit change request

File: sglang_omni/models/moss_tts_local/stages.py (hunk 6 above)
Comment (Ratish1) [reply sglang_omni/models/moss_tts_local/stages.py]: correct, getattr now uses only the explicit default

File: sglang_omni/models/moss_tts_local/stages.py (hunk 5 above)
Comment (Ratish1) [reply sglang_omni/models/moss_tts_local/stages.py]: done. moved the data-URI parse/decode helpers into _BatchedReferenceEncoder as private static methods.
